instrumentation/lettuce/lettuce-5.1/src/main/java/io/opentelemetry/javaagent/instrumentation/lettuce/v5_1/RedisCommandNormalizer.java

```diff
@@ -0,0 +1,208 @@
+    normalizers.put("ZREVRANGEBYSCORE", keepOneArg);
+    normalizers.put("ZREVRANK", keepOneArg);
+    normalizers.put("ZSCORE", keepOneArg);
+
+    // Streams
+    normalizers.put("XADD", new MultiKeyValue(2));
+
+    // Strings
+    normalizers.put("APPEND", keepOneArg);
+    normalizers.put("GETSET", keepOneArg);
+    normalizers.put("MSET", setMultiField);
+    normalizers.put("MSETNX", setMultiField);
+    normalizers.put("PSETEX", keepTwoArgs);
+    normalizers.put("SET", keepOneArg);
+    normalizers.put("SETEX", keepTwoArgs);
+    normalizers.put("SETNX", keepOneArg);
+    normalizers.put("SETRANGE", keepOneArg);
+
+    NORMALIZERS = unmodifiableMap(normalizers);
+  }
+
+  public static String normalize(String command, List<String> args) {
+    return NORMALIZERS
+        .getOrDefault(command.toUpperCase(), Default.INSTANCE)
+        .normalize(command, args);
+  }
+
+  public interface CommandNormalizer {
+    String normalize(String command, List<String> args);
+
+    enum Default implements CommandNormalizer {
+      INSTANCE;
+
+      @Override
+      public String normalize(String command, List<String> args) {
+        StringBuilder normalised = new StringBuilder(command);
+        for (String arg : args) {
+          normalised.append(" ").append(arg);
+        }
+        return normalised.toString();
+      }
+    }
+
+    class CommandAndNumArgs implements CommandNormalizer {
+      private final int numOfArgsToKeep;
+
+      public CommandAndNumArgs(int numOfArgsToKeep) {
+        this.numOfArgsToKeep = numOfArgsToKeep;
+      }
+
+      @Override
+      public String normalize(String command, List<String> args) {
+        StringBuilder normalised = new StringBuilder(command);
+        int i = 0;
+        for (; i < numOfArgsToKeep && i < args.size(); ++i) {
+          normalised.append(" ").append(args.get(i));
+        }
+        for (; i < args.size(); ++i) {
+          normalised.append(" ").append("?");
+        }
+        return normalised.toString();
+      }
+    }
+
+    class MultiKeyValue implements CommandNormalizer {
+      private final int numOfArgsBeforeKeyValue;
+
+      public MultiKeyValue(int numOfArgsBeforeKeyValue) {
+        this.numOfArgsBeforeKeyValue = numOfArgsBeforeKeyValue;
+      }
+
+      @Override
+      public String normalize(String command, List<String> args) {
+        StringBuilder normalised = new StringBuilder(command);
+        int i = 0;
+        // append all "initial" arguments before key-value pairs start
+        for (; i < numOfArgsBeforeKeyValue && i < args.size(); ++i) {
+          normalised.append(" ").append(args.get(i));
+        }
+
+        // whether keys are on even or odd index depends on the number of initial args
+        int keys = numOfArgsBeforeKeyValue % 2;
+        for (; i < args.size(); ++i) {
+          normalised
+              .append(" ")
+              // append only keys, skip values
+              .append(i % 2 == keys ? args.get(i) : "?");
+        }
```

Comment:
took me a while to understand the parity matching, I think something like this would be easier to read
```suggestion
        // loop over keys only
        for (i = numOfArgsBeforeKeyValue; i < args.size(); i += 2) {
          normalised
              .append(" ")
              .append(args.get(i))
              .append(" ?");
        }
```

Reply:
That's much better!